src/Tasks/ResourceHandling/MSBuildResXReader.cs

```diff
@@ -0,0 +1,299 @@
+                return;
+            }
+            else
+            {
+                switch (mimetype)
+                {
+                    case ByteArraySerializedObjectMimeType:
+                        // TypeConverter from byte array
+                        byte[] typeConverterBytes = Convert.FromBase64String(value);
+
+                        resources.Add(new TypeConverterByteArrayResource(name, typename, typeConverterBytes, resxFilename));
+                        return;
+                    case BinSerializedObjectMimeType:
+                    case Beta2CompatSerializedObjectMimeType:
+                    case CompatBinSerializedObjectMimeType:
+                        // BinaryFormatter from byte array
+                        byte[] binaryFormatterBytes = Convert.FromBase64String(value);
+
+                        resources.Add(new BinaryFormatterByteArrayResource(name, binaryFormatterBytes, resxFilename));
+                        return;
+                    default:
+                        throw new NotImplementedException();
+                }
+            }
+        }
+
+        private static void AddLinkedResource(string resxFilename, bool pathsRelativeToBasePath, List<IResource> resources, string name, string value)
+        {
+            string[] fileRefInfo = ParseResxFileRefString(value);
+
+            string fileName = FileUtilities.FixFilePath(fileRefInfo[0]);
+            string fileRefType = fileRefInfo[1];
+
+            if (pathsRelativeToBasePath)
+            {
+                fileName = Path.Combine(
+                    FileUtilities.GetDirectory(
+                        FileUtilities.NormalizePath(resxFilename)),
+                    fileName);
+            }
+
+            if (fileRefType == StringTypeName)
+            {
+                string fileRefEncoding = null;
+                if (fileRefInfo.Length == 3)
+                {
+                    fileRefEncoding = fileRefInfo[2];
+
+#if RUNTIME_TYPE_NETCORE
+                    // Ensure that all Windows codepages are available.
+                    // Safe to call multiple times per https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.registerprovider
+                    Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
+#endif
+                }
+
+                // from https://github.com/dotnet/winforms/blob/a88c1a73fd7298b0a5c45251771f439262016826/src/System.Windows.Forms/src/System/Resources/ResXFileRef.cs#L231-L241
+                Encoding textFileEncoding = fileRefEncoding != null
+                    ? Encoding.GetEncoding(fileRefEncoding)
+                    : Encoding.Default;
+                using (StreamReader sr = new StreamReader(fileName, textFileEncoding))
+                {
+                    resources.Add(new StringResource(name, sr.ReadToEnd(), resxFilename));
+
+                    return;
+                }
+            }
+
```

Comment:
I think you need a special case for `.ico` here, per https://github.com/dotnet/winforms/blob/33b9fe202f3dc1b8e7c4bf28492f8bd3252f1a20/src/System.Windows.Forms/src/System/Resources/ResXFileRef.cs#L309-L315

Reply:
Thinking about this was on my to-do list, but didn't make it into the checklist above.

If I'm reading that correctly, wouldn't it apply only at _deserialization_ time? The thing that got serialized into `.resources` before was a `System.Drawing.Bitmap`; it just got created via a two-hop process instead of just calling `Activator.CreateInstance` with the stream. I don't see an analogue for that in the new world. Am I missing something?

Reply:
You change the type name you pass in.  Instead of passing in System.Drawing.Bitmap, you need to pass in System.Drawing.Icon.  We can't make the decision at deserialization time because we don't know what the file extension was.

Reply:
I keep struggling with this, both

## How to create this situation to test

Is there a VS UI way to get this "Bitmap read from an `.ico` file" situation? I wasn't able to find one--references to icons get returned as `System.Drawing.Icon` and/or embedded, not file-refed.

I can make it manually. If I do, do we know how to validate the reported GDI+ mangling?

## What the correct behavior is

The way I'm interpreting the current behavior, it applies _only_ when you're in this kind of situation

```xml
  <data name="pictureBox1.Image" type="System.Resources.ResXFileRef, System.Windows.Forms">
    <value>test.ico;System.Drawing.Bitmap, System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</value>
  </data>
```

and what you get out at runtime is a `Bitmap`, it just has different bytes than the `Bitmap` that would have been loaded using the normal codepath, because it bounced through `Icon`. Right? I'm not seeing how we could produce that other than a special case in `System.Resources.Extensions`, something like a bool `BounceBitmapThroughIcon` at deserialize time.

Reply:
> Is there a VS UI way to get this "Bitmap read from an .ico file" situation? 

What if you have a Bitmap localizable property on a control and you point it to an Icon file?  I see lots of [hits](https://github.com/search?q=%22.ico%3BSystem.Drawing.Bitmap%22&type=Code) on this pattern so there must be something that produces it.  Winforms also has a test case: https://github.com/dotnet/winforms/blob/f753f5bad925233c65a522921f9eeaf2b46b9c78/src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXFileRef.ConverterTests.cs#L111-L122.

> do we know how to validate the reported GDI+ mangling?

Double check that alpha channel remains transparent.  So use an icon with some transparency then put it on top of some colored background and make sure it stays transparent rather than defaulting to some color.

> like a bool BounceBitmapThroughIcon at deserialize time.

Hmmm, that's unfortunate.  Creating API for special cases like this is not a great solution.  It's also less than ideal to store all the images in the ico when only one will ever be exposed by the app.  At the moment I can't think of a great alternative.  We could emit a suppressible warning when we see this pattern that recommends that folks export the icon to an image...  Alternatively the task could do this special casing.

Reply:
> What if you have a Bitmap localizable property on a control and you point it to an Icon file?

Ah, I didn't try with a localizable property.

> Alternatively the task could do this special casing.

 This feels like the best bad option at this point.

> use an icon with some transparency then put it on top of some colored background and make sure it stays transparent rather than defaulting to some color.

Trying this: https://github.com/rainersigwald/resx-experiments/tree/transparent-fileref-icon/transparent-fileref-icon, with full framework serialization:
![image](https://user-images.githubusercontent.com/3347530/59289442-b6d6cb80-8c3b-11e9-9e92-b06bdaa439d0.png)

with no special cases on top of this PR:
![image](https://user-images.githubusercontent.com/3347530/59289533-eb4a8780-8c3b-11e9-9d3b-a092e4a06702.png)

with special casing to change the type to Icon from Bitmap:
![image](https://user-images.githubusercontent.com/3347530/59289362-8000b580-8c3b-11e9-88fa-9ab28d3df66d.png)

with special casing to read an icon from the file and serialize `ToBitmap` in the legacy live-object way:

Reply:
Those diffs appear to me like Icon.ToBitmap behaves differently on Core.  Looks like it is selecting the wrong bitmap, or is generating one with wrong bit depth. 

Interestingly it seems like alpha channel is missing from .net framework case: I wonder if you have to configure that control to honor alpha.

Reply:
Nope, I lied: the special-cased final one isn't doing the right thing. Fixing.

Reply:
Sorry, that was confusing: the background of the window is lime; anything with lime showing through is handling alpha correctly. The pixels of the icon itself are white+transparent, not lime+transparent.

Reply:
We talked about this offline. There are a few options, but anything that involves getting a live System.Drawing.Bitmap during the build is awkward, because not every .NET Core installation on Linux or macOS will have libgdiplus, which is needed for bitmaps.

For now, we're going to elide this special case. The reported transparency mangling may have been fixed (see images above: lime green is the background color so at least some alpha is being preserved). If this causes problems, the decision can be revisited.